rust/arrow/src/array/array.rs

```diff
@@ -1647,6 +1669,172 @@ impl From<(Vec<(Field, ArrayRef)>, Buffer, usize)> for StructArray {
     }
 }
 
+/// A dictonary array where each element is a single value indexed by an integer key.
+/// This is mostly used to represent strings or a limited set of primitive types as integers,
+/// for example when doing NLP analysis or representing chromosomes by name.
+pub struct DictionaryArray<K: ArrowPrimitiveType> {
+    // Array of keys, much like a PrimitiveArray
+    data: ArrayDataRef,
+
+    // Pointer to the key values.
+    raw_values: RawPtrBox<K::Native>,
+
+    // Array of any values.
+    values: ArrayRef,
+
+    /// Values are ordered.
+    is_ordered: bool,
+}
+
+pub struct NullableIter<'a, T> {
+    data: &'a ArrayDataRef,
+    keys: &'a [T],
+    i: usize,
+}
+
+impl<'a, T> std::iter::Iterator for NullableIter<'a, T>
+where
+    T: Clone,
+{
+    type Item = Option<T>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let i = self.i;
+        if i >= self.keys.len() {
+            None
+        } else if self.data.is_null(i) {
+            self.i += 1;
+            Some(None)
+        } else {
+            self.i += 1;
+            Some(Some(self.keys[i].clone()))
+        }
+    }
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        (self.keys.len(), Some(self.keys.len()))
+    }
+
+    fn nth(&mut self, n: usize) -> Option<Self::Item> {
+        let i = self.i;
+        if i + n >= self.keys.len() {
+            self.i = self.keys.len();
+            None
+        } else if self.data.is_null(i + n) {
+            self.i += n + 1;
+            Some(None)
+        } else {
+            self.i += n + 1;
+            Some(Some(self.keys[i + n].clone()))
+        }
+    }
+}
+
+impl<'a, K: ArrowPrimitiveType> DictionaryArray<K> {
+    /// Returns a slice of the keys (offsets into the values array).
+    /// Note that if nulls exist, you must check them.
+    pub fn keys(&self) -> &[K::Native] {
```

Comment:
I'll think about it over the weekend. I did have the DictionaryArray as two separate child arrays and we could give the ArrayData two children instead of one.